src/App.js

```diff
@@ -5,110 +5,84 @@ class App extends Component {
   constructor(props) {
     super(props);
     this.statuses = ['not yet signed in', 'signed in', 'error message'];      // bad practice?
-    this.default_states = {
+    this.defaultStates = {
       status: this.statuses[0],
       user: "",
       profile: ""
     };
-    this.state = this.default_states;
+    this.state = this.defaultStates;
 
-    this.onLogin = this.onLogin.bind(this);
-    this.onLogout = this.onLogout.bind(this);
-    this.onError = this.onError.bind(this);
-  }
-
-  onLogin() {
-    skygear.auth.loginOAuthProviderWithPopup('google').then(
-      // sign in was successful
-      user => {
-        console.info('Login success', user);
-        console.info('Access token:', skygear.auth.accessToken);
-        console.info('Username:', skygear.auth.currentUser.username);
-
-        // status becomes signed in
-        this.setState({ status: this.statuses[1] });
-        skygear.auth.whoami().then(
-          // whoami returns a Record type
-          record => this.setState({ user: JSON.stringify(record.toJSON(), null, 2) }),
-          this.onError('')
-        );
-        skygear.auth.getOAuthProviderProfiles().then(
-          // getOAuthProviderProfiles return a JSON type
-          profileJson => this.setState({ profile: JSON.stringify(profileJson, null, 2) }),
-          this.onError('')
-        );  
-      },
-      // sign in was unsuccessful
-      this.onError('Login failure')
-    );
-  }
+    this.onLogin = () => {
+      skygear.auth.loginOAuthProviderWithPopup('google')
+        .then((value) => {
+          // value correponds to loginOAuthProviderWithPopup returning user
+          this.setState({ user: JSON.stringify(value.toJSON(), null, 2) });
+          // after loginOAuthProviderWithPopup, we can legally getOAuthProviderProfiles
+          skygear.auth.getOAuthProviderProfiles()
+            .then((value) => {
+              // value correponds to getOAuthProviderProfiles returning profileJson
+              this.setState({ profile: JSON.stringify(value, null, 2) });
+            }).catch(this.onError('Login failure at getOAuthProviderProfiles'));  
+        }).then(() => {
+          // sign in was successful
+          console.info('Login success');
+          console.info('Access token:', skygear.auth.accessToken);
+          console.info('Username:', skygear.auth.currentUser.username);
+          // status becomes signed in
+          this.setState({ status: this.statuses[1] });
+        }).catch(this.onError('Login failure at loginOAuthProviderWithPopup'));
+    };
```

Comment:
You can run promises in series with:

```js
skygear.auth.loginOAuthProviderWithPopup(xxx)
  .then((xxx) => {
     // xxxx
    return skygear.auth. getOAuthProviderProfiles();
  }).then((xxx) => {
    // xxx
  })
  .catch((xxx) => {
    // xxx;
  });
```

Reply:
If you don't return a promise in the `then` function, the next `then` function would be executed immediately.

So your updated version runs into the same problem of the old code.